### ISSUE TITLE
Example config for Database Session Persistence feature

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#1289]
Example config for the `sessionDatabase-1.0` feature: https://openliberty.io/docs/ref/feature/#sessionDatabase-1.0.html

**Resources**:
- https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_session_persistence.html

It might be worth searching the KC for other examples that you can dig up before reaching out to an SME.

**Comment by Rwalls1** [comment]
@ManasiGandhi The topic looks good, I just have a few comments: 
- Make sure to run Acrolinx again as I noticed some things that were flagged
- Add a period after, "Failover of HTTP sessions can be achieved by configuring multiple servers to persist data to the same location” ﻿
- For the paragraph that begins with, "After enabling the sessionDatabase feature, define a data source, `<dataSource id="SessionDS" …​ />﻿`”, and ends with "Then refer...".  I think it reads a bit awkwardly.  I think you should avoid phrasing as a command here and instead change "define" to "you can define" or something like that as these are examples and not a task. Also, if presented this way,  I think the rest of the paragraph should be broken up so the references to the config examples are clearer and more consistent. So maybe change the paragraph to something like: 
"After you enable the `sessionDatabase` feature, you can define a data source, as shown in the following example:
`<dataSource id="SessionDS" …​ />`
When you define a data source, you can refer to the data source from the following session database configuration:
`<httpSessionDatabase id="SessionDB" dataSourceRef="SessionDS" …​ />`
You can also refer to the persistent storage location from the following session management configuration:
<httpSession storageRef="SessionDB" …​ />"
- For the phrase, "The example creates a file named ${shared.config.dir}/httpSessionPersistence.xml:", change to "The following example..."
- There are other instances throughout the topic where you phrase sentences as commands that I think should be avoided.

**Comment by zhang-ca** [comment]
https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#sessionDatabase-1.0.html

Reviewed the draft, looks good. Also suggest to add :
`See example in bootstrap.properties files. `
after
`The clone ID must be unique for each server and can be 8-9 alphanumeric characters in length.`

**Comment by zhang-ca** [comment]
`as shown in the example of bootstrap.properties file.` doesn't look too well on top of server.xml example.

I suggest to move the following paragraphs on top of **You can specify unique variables in the bootstrap.properties file of each server.**

`The HTTP server plugin uses the clone ID that is inserted into the response/request header to maintain session affinity between requests. In Open Liberty, the clone ID is generated when you start a server for the first time and it is regenerated if you start the server with the --clean option. For production use, manually assigning a clone ID will ensure that the ID is stable and that request affinity is correctly maintained. The clone ID must be unique for each server and can be 8-9 alphanumeric characters in length.`

**Comment by zhang-ca** [comment]
Looks good, nice and clean.

**Comment by lauracowen** [comment]
The format of these topics is not to provide step-by-step (task-style) instructions for configuring something. It's simply to show one or more commonly needed examples of configuration. Those examples must not show the whole `server.xml`content either. If there is a common need for a set of step-by-step instructions for configuring a technology on Open Liberty, raise an issue on the docs backlog separately and explain who needs it and why it's needed. See [How to add config examples to generated feature docs](https://ibm.box.com/s/v3oejxfcv1e63me9hxgot3qdkjrmjby4).

- The first example should be the most basic config, for a single server only. The heading should clearly describe the example, not be phrased as if it's a generic task topic.
- If other examples are also needed (eg for multiple servers persisting to the same database - if this is a common configuration to use?), create a second sub-heading with its own example. Again, though, include only the relevant parts of the server.xml, not the whole thing or standard app configuration lines.
- Remove the last part about variables in the bootstrap.properties file. It's not relevant here. Instead, in examples, use hard-coded values then explain what those values do.

@zhang-ca  Please can you clarify what parts of the example server.xml are specific to this feature and what is the _minimum_ configuration that would be needed for session persistence to work?

**Comment by zhang-ca** [comment]
For a single server example of server.xml, here is the minimum configuration that would be needed for session persistence to work. The feature is specified in  `<feature>sessionDatabase-1.0</feature>`

```
<server description="Demonstrates HTTP Session Persistence Configuration">

    <featureManager>
        <feature>sessionDatabase-1.0</feature>
        <feature>servlet-3.0</feature>
    </featureManager>

    <fileset id="DerbyFiles" includes="*.jar" dir="${shared.resource.dir}/derby/client"/>
    <library id="DerbyLib" filesetRef="DerbyFiles"/>
    <jdbcDriver id="DerbyDriver" libraryRef="DerbyLib"/>
    <dataSource id="SessionDS" jdbcDriverRef="DerbyDriver">
        <properties.derby.client user="user1" password="password1"
                                 databaseName="${shared.resource.dir}/databases/SessionDB"
                                 createDatabase="create"/>
    </dataSource>

    <httpSessionDatabase id="SessionDB" dataSourceRef="SessionDS"/>

</server>
```

**Comment by lauracowen** [comment]
Okay, thanks. So in the example in the topic, we'd show and describe:

```
    <fileset id="DerbyFiles" includes="*.jar" dir="${shared.resource.dir}/derby/client"/>
    <library id="DerbyLib" filesetRef="DerbyFiles"/>
    <jdbcDriver id="DerbyDriver" libraryRef="DerbyLib"/>
    <dataSource id="SessionDS" jdbcDriverRef="DerbyDriver">
        <properties.derby.client user="user1" password="password1"
                                 databaseName="${shared.resource.dir}/databases/SessionDB"
                                 createDatabase="create"/>
    </dataSource>

    <httpSessionDatabase id="SessionDB" dataSourceRef="SessionDS"/>
```

(we wouldn't show all the feature enablement bit as that's been covered in the previous section)

**Comment by ManasiGandhi** [comment]
@Charlotte-Holt Can you peer review this draft https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#sessionDatabase-1.0.html?

**Comment by zhang-ca** [comment]
The dataSourceRef **attribute** points to the dataSource element with the id attribute SessionDS. The jdbcDriverRef **attribute** points to the jdbcDriver element with the id attribute DerbyDriver.

Include the shared session management configuration in each of your servers. For example, create two server.xml files for server instances named s1 and s2, as follows:

    ${wlp.user.dir}/servers/s1/server.xml
    ${wlp.user.dir}/servers/s2/server.xml

<server description="Example Server">
    <include location="${shared.config.dir}/httpSessionPersistence.xml"/>
</server>

**Comment by lauracowen** [comment]
- the format looks cleaner now, thanks.
- [ ] Don't start with "After you enable the xxx feature" as that goes without saying - it's just redundant in these examples.
- [ ] Don't talk about the `httpSessionPersistence.xml` file or including files. Just intro the example as showing the minimum configuration for database persistence. Don't explain how to do the includes etc.
- [ ] Change the heading to be more explanatory: eg "Minimum configuration to persist session data to a database" (instead of "Configure Liberty server...etc").
- [ ] In the text after the example, apply Allan's comment "The dataSourceRef **attribute** points to the dataSource element with the id attribute SessionDS. The jdbcDriverRef **attribute** points to the jdbcDriver element with the id attribute DerbyDriver."

- [ ] If Allan feels it's necessary to more explicitly explain how to do the includes etc, raise a separate issue on the docs backlog.

- [ ] Can you also make the directory structure use the consistent format as described here: https://github.com/OpenLiberty/docs/issues/1765

**Comment by zhang-ca** [comment]
If Don't talk about the httpSessionPersistence.xml, the include file I mentioned above is not required.

**Comment by lauracowen** [comment]
Thanks signing off.

**Comment by ManasiGandhi** [comment]
@chirp1 updated draft link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/sessionDatabase-1.0.html

**Comment by dmuelle** [comment]
need to replace `${server.config.dir}` variable with an example (use `server1` for consistency ) per #1877. Once the file is updated, please mark it off from the checklist in that issue- thanks!

**Comment by ManasiGandhi** [comment]
@dmuelle I replaced ${server.config.dir} variable with an example with server1 in [#1289](https://github.com/openliberty/docs/issues/1289).

**Comment by ManasiGandhi** [comment]
New link for the draft  https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/sessionDatabase-1.0.html

**Comment by ManasiGandhi** [comment]
@chirp1 updated link to the draft https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/feature/sessionDatabase-1.0.html.

**Comment by ManasiGandhi** [comment]
@chirp1 20.12 updated link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/sessionDatabase-1.0.html

**Comment by zhang-ca** [comment]
It would be great if we can line up the indentation below: (i.e. line up **fileset** with **library**)

>     <fileset id="DerbyFiles" includes="*.jar" dir="server1/derby/client"/>
>     <library id="DerbyLib" filesetRef="DerbyFiles"/>
>     <jdbcDriver id="DerbyDriver" libraryRef="DerbyLib"/>
>     <dataSource id="SessionDS" jdbcDriverRef="DerbyDriver">
>         <properties.derby.client user="user1" password="password1"
>                                  databaseName="server1/databases/SessionDB"
>                                  createDatabase="create"/>
>     </dataSource>
> 
>     <httpSessionDatabase id="SessionDB" dataSourceRef="SessionDS"/>

**Comment by lauracowen** [comment]
@ManasiGandhi I'm not sure what's happened to the indentation (either I missed it when I reviewed, or something's changed since I reviewed it). Can you fix this indentation as @zhang-ca requests above? Also, that snippet seems to be marked as "Java" source code but it's not, it's XML and should be marked as that so that the syntax highlighting is correct.

Also "SessionDS" and "DerbyDriver" should be highlighted as values.

**Comment by ManasiGandhi** [comment]
@lauracowen @zhang-ca I worked on the review

- [x] Can you fix this indentation as @zhang-ca requests above? 

- [x] Also, that snippet seems to be marked as "Java" source code but it's not, it's XML and should be marked as that so that the syntax highlighting is correct.

**Comment by lauracowen** [comment]
Hi, I don't see the changes on the draft server:

![image.png](https://images.zenhubusercontent.com/583eaf59b9dc3c622022b986/c8f3c750-8cbb-4167-8020-1726dbb3c1ae)

**Comment by ManasiGandhi** [comment]
I worked on Laura's review for this issue and waiting on the changes to show after the builds are resolved.

**Comment by ManasiGandhi** [comment]
Laura's comments have been addressed.

**Comment by ManasiGandhi** [comment]
@rw2513 can you peer review this topic.

**Comment by Rwalls1** [comment]
## Peer review

@ManasiGandhi The example looks good, I just have a couple comments:
- In the example, I think the indentation still needs to be fixed as Allan and Laura requested.
- While I think you’ve done a good job of highlighting an example of the minimum configuration to persist session data and explaining the basic components, I think the example still needs a bit more framing to fully capture its significance. In addition to explaining what’s happening in the example, I think you should also explain the function of the components that you highlighted. So, instead of just saying, "The `dataSourceRef` attribute points to the `dataSource` element”, or "The `jdbcDriverRef` attribute points to the `jdbcDriver` element”, I think you should also briefly explain why that’s important and/or how that helps to successfully configure database persistence. I think adding that additional context will make the example more helpful and easier to understand.

**Comment by ManasiGandhi** [comment]
@Rwalls1 I worked on your peer review https://draft-openlibertyio.mybluemix.net/docs/21.0.0.2/reference/feature/sessionDatabase-1.0.html.

- [x] In the example, I think the indentation still needs to be fixed as Allan and Laura requested.
While I think you’ve done a good job of highlighting an example of the minimum configuration to persist session data and explaining the basic components, I think the example still needs a bit more framing to fully capture its significance. 

- [x] In addition to explaining what’s happening in the example, I think you should also explain the function of the components that you highlighted. So, instead of just saying, "The dataSourceRef attribute points to the dataSource element”, or "The jdbcDriverRef attribute points to the jdbcDriver element”, I think you should also briefly explain why that’s important and/or how that helps to successfully configure database persistence. I think adding that additional context will make the example more helpful and easier to understand

**Comment by dmuelle** [comment]
exmaples no longer in use, superseded by content in the
